runtime/src/iree/hal/drivers/hip/hip_allocator.c

```diff
@@ -697,8 +773,36 @@ iree_status_t iree_hal_hip_allocator_free_async(
       &allocator->statistics, iree_hal_buffer_memory_type(buffer),
       iree_hal_buffer_allocation_size(buffer)));
 
-  IREE_RETURN_IF_ERROR(IREE_HIP_CALL_TO_STATUS(allocator->symbols,
-                                               hipFree(device_ptr), "hipFree"));
+  iree_status_t status = iree_ok_status();
+  iree_slim_mutex_lock(&allocator->async_allocation_mutex);
+  iree_hal_hip_util_tree_node_t* bucket = iree_hal_hip_util_tree_get(
+      &allocator->async_allocation_map.tree,
+      (iree_host_size_t)iree_hal_buffer_allocation_size(buffer));
+  if (bucket == NULL) {
+    status = iree_hal_hip_util_tree_insert(
+        &allocator->async_allocation_map.tree,
+        (iree_host_size_t)iree_hal_buffer_allocation_size(buffer), &bucket);
+    iree_hal_hip_async_allocation_map_item_t* queue_item =
+        (iree_hal_hip_async_allocation_map_item_t*)
+            iree_hal_hip_util_tree_node_get_value(bucket);
+    iree_hal_hip_async_allocation_queue_initialize(allocator->host_allocator,
+                                                   &queue_item->queue);
+  }
+  if (iree_status_is_ok(status)) {
+    iree_hal_hip_async_allocation_t allocation = {
+        .allocation_size = iree_hal_buffer_allocation_size(buffer),
+        .pointer = device_ptr};
+    iree_hal_hip_async_allocation_queue_push_back(
+        &((iree_hal_hip_async_allocation_map_item_t*)
+              iree_hal_hip_util_tree_node_get_value(bucket))
+             ->queue,
+        allocation);
+  }
+  iree_slim_mutex_unlock(&allocator->async_allocation_mutex);
+
+  if (!iree_status_is_ok(status)) {
```

Comment:
as mentioned in other PRs don't exit early unless you need to:

```
if (iree_status_is_ok(status)) {
  iree_hal_hip_buffer_set_allocation_empty(buffer);
}
return status;
```

there should be very few cases where you need to `if !ok return`

doing this makes it clearer what is a success case vs not, defines a structured place for cleanup (which this may need), and makes inserting tracing easier (as there's a clear place to put the end zone).

in cases where iree_hal_hip_util_tree_insert succeeds but iree_hal_hip_async_allocation_queue_push_back fails you've still got the entry bucket in the tree - is that expected?